Service/src/main/java/org/apidb/apicommon/service/services/jbrowse/JBrowseUserDatasetsService.java

```diff
@@ -1,161 +1,206 @@
-import org.gusdb.wdk.service.formatter.UserDatasetsFormatter;
-import org.gusdb.wdk.service.service.user.UserDatasetService;
+import javax.ws.rs.core.Response;
+
 import org.gusdb.wdk.service.service.user.UserService;
-import org.json.JSONArray;
-import org.json.JSONObject;
+
+import java.io.File;
+import java.nio.file.Paths;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.gusdb.wdk.service.FileRanges.getFileChunkResponse;
+import static org.gusdb.wdk.service.FileRanges.parseRangeHeaderValue;
 
 public class JBrowseUserDatasetsService extends UserService {
 
   private static final Logger LOG = Logger.getLogger(JBrowseUserDatasetsService.class);
+  private static final String VDI_DATASET_DIR_KEY = "VDI_DATASETS_DIRECTORY";
+  private static final String VDI_CONTROL_SCHEMA_KEY ="VDI_CONTROL_SCHEMA";
+  public static final String BIGWIG_SUBCATEGORY = "Bigwig Files From User";
+  public static final String RNA_SEQ_SUBCATEOGRY = "RNASeq";
 
   public JBrowseUserDatasetsService(@PathParam(USER_ID_PATH_PARAM) String uid) {
     super(uid);
   }
 
+
+  /**
+   * This endpoint exposes user dataset track files with the intent of JBrowse using it to retrieve file ranges.
+   * Any installed file for a dataset owned by the user can be fetched by this endpoint.
+   *
+   * URLs to hit this endpoint are constructed by this service when the JBrowse application asks for the user's available
+   * tracks.
+   */
   @GET
-  @Path("user-datasets-jbrowse/{organism}")
+  @Path("user-datasets-jbrowse/data")
   @Produces(MediaType.APPLICATION_JSON)
-  public JSONObject getAllUserDatasetsJBrowse(@PathParam("organism") String publicOrganismAbbrev) {
+  public Response getAllUserDatasetFileJBrowse(@QueryParam("data") String data,
+                                               @QueryParam("datasetID") String datasetID,
+                                               @HeaderParam("Range") String fileRange) throws WdkModelException {
+    String buildNumber = getWdkModel().getBuildNumber();
+    String udDir = getWdkModel().getProperties().get("VDI_DATASETS_DIRECTORY");
+
+    // Verify that the dataset belongs to user. Random people should not be able to download anyone's files, even
+    // though they are protected by obscurity.
+    datasetBelongsToUser(getPrivateRegisteredUser().getUserId(), datasetID);
+
+    String path = String.format("%s/build-%s/%s/%s/%s", udDir, buildNumber, getWdkModel().getProjectId(), datasetID, data);
+
+    if (path.contains("..") || path.contains("$")) {
+      throw new NotFoundException(formatNotFound("*"));
+    }
+
+    return getFileChunkResponse(Paths.get(path), parseRangeHeaderValue(fileRange));
+  }
+
 
+  @GET
+  @Path("user-datasets-jbrowse/{organism}")
```

Comment:
Looks like you're no longer using the organism param.  True?  If so should maybe get rid of it.  Don't know how this affects the UI though (API change)

Reply:
True! We no longer need to check the dependencies here since it's done at install time. I might make a separate issue for this to keep the initial change compatible with the current UI.